sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -748,23 +751,74 @@ function getOrCreateEntry(
     dispatchAny();
     return entry;
 
-    function dispatchAny(): void {
+    function doNotCapture() {
         if (!serialize(obj)) {
             // caller explicitly does not want us to capture this value.
-            entry.json = undefined;
+            return true;
         }
-        else if (obj && obj.doNotCapture) {
+
+        if (obj && obj.doNotCapture) {
             // object has set itself as something that should not be captured.
-            entry.json = undefined;
+            return true;
         }
-        else if (isDerivedNoCaptureConstructor(obj)) {
+
+        if (isDerivedNoCaptureConstructor(obj)) {
             // this was a constructor that derived from something that should not be captured.
+            return true;
+        }
+
+        return false;
+    }
+
+    function dispatchAny(): void {
+        if (doNotCapture()) {
+            // We do not want to capture this object.  Explicit set .json to undefined so
+            // that we will see that the property is set and we will simply roundtrip this
+            // as the 'undefined value.
             entry.json = undefined;
+            return;
         }
-        else if (obj === undefined || obj === null ||
+
+        if (obj === undefined || obj === null ||
             typeof obj === "boolean" || typeof obj === "number" || typeof obj === "string") {
             // Serialize primitives as-is.
             entry.json = obj;
+            return;
+        }
+
+        const moduleName = findModuleName(obj);
+        if (moduleName && moduleName[0] !== ".") {
+            // @pulumi modules can't ever be referenced on the 'inside'.  Not only are they not
+            // written to support being used on the inside, but serialization also explicitly
+            // removes these modules from node_modules to trim down the total upload size.
+            //
+            // If the user's code actually explicitly references this, then emit a hard error
+            // to let them know what's going on.
+            if (moduleName.startsWith("@pulumi")) {
+                // Find the 'func' to report this issue against. Note: we will always find some
+                // 'func' in the context chain as the only reason we ever even get into closure
+                // creation is because we're serializing out some function.
+                let func: Function = <any>undefined;
+                for (let i = context.frames.length - 1; i >= 0; i--) {
+                    const frame = context.frames[i];
+                    if (frame.functionLocation) {
+                        func = frame.functionLocation.func;
+                    }
+                }
+
+                throwSerializationError(func, context,
```

Comment:
This seems semi-arbitrary to me; can you remind me why we need to do this? Firstly, virtually anything can be put into a @pulumi module, so it's not clear that by definition eliding them won't lead someone to be in a situation where they can't access important functionality (like a helper function). Secondly, hard coded rules like this always frustrate me as a user.

If it's just that gRPC is too large, why don't we filter out native modules, or something more generalizable and principled
?

Reply:
> This seems semi-arbitrary to me; can you remind me why we need to do this?

For one, because we explicitly do not upload any @pulumi node_modules.   So if we actually capture a pulumi-module *by* reference, then that will def break at runtime. :)

This is both because that node_module is enormous, and because much of it does not work on the 'inside' due to needing to be able to do things like speak to the pulumi engine.

> so it's not clear that by definition eliding them won't lead someone to be in a situation where they can't access important functionality (like a helper function). 

Note: using a helper function is fine.  In essence, that's what is happening when you use "pulumi.Config".  If you actually use anything *in* the module, then that is captured by value 'as expected', and it serialized properly into your lambda code.  

What is being prevented here is literally being able to actually use the module-object itself.  

> If it's just that gRPC is too large, why don't we filter out native modules, 

I'm not sure what counts as a native module :)   We definitely don't want to prevent a module that *uses* anything native.  in practice, most modules will end up doing that in some manner.  

> or something more generalizable and principled?

I think the problem is that it's really hard to find an actual 'principle' that doesn't have some problem.  For example, our old principle was "do not special case modules.  capture everything by value."  That was principled, but broke down badly with modules (forcing people to do things like 'require on the inside').  Similarly, we tried the principle of capturing all modules by by 'import'.  But this ends up breaking user-code itself because now we don't actually properly serialize code that can work properly since our serialization strategy does not properly mimic the user's actual code layout (and trying to do so would be a large undertaking).

This approach tries to accomplish the following:

1. allow users to write the code in idiomatic TS/JS, and have the semantics for all their module code behave as expected.
2. have a capturing story that has simple and explainable semantics.
3. allow people to author  a single project.json that defines both the deployment-time and cloud-runtime-time configuration, with reasonable outputs for both environments.

After this PR we have the above.  However, there is one specific issue a user could potentially trip themselves into.  Namely, they end up wanting to actually capture a pulumi module reference as a reference.  

Right now we can't think of any good reason why someone would want to do this.  And, instead of having it silently blow up, i thought it would be good to at least have us let the user know they're doing something bad.  

Note: In the future, if we think there is a good reason for this.  We can definitely relax this restriction.  But i'd rather start with a good working system, with a single non-onerous restriction, and only relax later if we really think it's necessary.

Reply:
@lukehoban Would def appreciate your thoughts here.

Reply:
> This is both because that node_module is enormous

But, why is it enormous? And why is an enormous package specific in any way to `@pulumi`?

> much of it does not work on the 'inside' due to needing to be able to do things like speak to the pulumi engine.

Yes, but the key word in here is _much_. There is some of it that could work.

I understand that the pragmatic, expedient thing here is to just filter out `@pulumi`, but it really doesn't feel like this is where we want to land long-term.

This PR is a great step forward, so I'm not blocking this change, but any magic name-based special casing is always a big code smell to me, and this part has me a little concerned.

Reply:
> But, why is it enormous? 

The primary reasons it is enormous is because pulumi pulls in both grpc (56MB) and typescript (35mb).  Together, we're already close to 100MB on just these two packages.  These two packages are only needed by pulumi at deployment time to:

1. Use GRPC to communicate with teh pulumi engine (something not needed at cloud-runtime). 
2. Use TypeScript ot analyzer serialize a javascript function into the code needed by the AWS lambda (something not needed at cloud-runtime).

> And why is an enormous package specific in any way to @pulumi?

It's specific to pulumi because the @pulumi libraires are the ones that are so specifically intended primarily for deployment time. This is different form (literally) every other library out there, which very much *is* intended to be used at runtime by an application. 

It's also specific to @pulumi in that users have no control here about including those packages.  You have to include those packages, or else pulumi simply doesn't work.  However, any other packages you don't need, you can simply leave out of your package.json.  However, Pulumi Applications are themselves special in that they have two very distinct lifetimes.  The local deployment lifetime, and the remote cloud lifetime.  @pulumi is needed for the former, but we don't want the usage int he former to cost you for the latter. :)

> Yes, but the key word in here is much. There is some of it that could work.

Note: it's important to be very precise about what does or doesn't work here.  If you end up trying to use part of pulumi which *can* run in the cloud then it *will* work.  That's part of this work.  The poster-child for this is "pulumi.Config".  You can use Config inside the cloud because it is specifically accessing the part of the pulumi that will work, while side stepping the parts that will not work. 

Similarly, you can obviously capture and use any sort of pulumi resource (even though those came from an @pulumi module).  What does *not* work is trying to write code that effectively would become a `require("@pulumi")` at runtime.  So far, there are no scenarios we can envision where could would want to actually do that.

Reply:
> This PR is a great step forward, so I'm not blocking this change, but any magic name-based special casing is always a big code smell to me, and this part has me a little concerned.

Sounds good to me.  Note that IMO, it's much easier to start with carving out things that we do not want to support up front.  Later on, we can relax in a conscientious and deliberate fashion based on very intentional and specific scenarios to support.

Reply:
I do think it would be good to open the issue tracking adding a first class marker for this instead of relying on `@pulumi/` scope.  And then link that issue as `TODO` here in the code where we hard code `@pulumi/`.

Reply:
Having a `@deploymentOnly()` marker would make me feel better. The reason I grow concerned when I see `@pulumi` hard-coded is (1) there are potentially libraries that would be available in `@pulumi` packages that are _not_ just deployment-time, and (2) there are potentially libraries outside of `@pulumi` (such as 3rd party ones) that _are_ only deployment-time.

Note that if we only did this for `@pulumi/pulumi`, our runtime SDK, I'd not have felt the same way.

Reply:
https://github.com/pulumi/pulumi/issues/1676 Tracks this.  Note: my preference would be that work here be driven by a real use case that we can use to vet any solution we create.